beamer/state_machine.py

```diff
@@ -241,9 +242,18 @@ def _handle_request_resolved(event: RequestResolved, context: Context) -> Handle
     return True, None
 
 
-def _handle_fill_hash_invalidated(_event: FillHashInvalidated, _context: Context) -> HandlerResult:
-    # TODO: implement once contract is finished
-    return False, None
+def _handle_fill_hash_invalidated(event: FillHashInvalidated, context: Context) -> HandlerResult:
+    request = context.requests.get(event.request_id)
+    if request is None:
+        return False, None
+
+    # Mark the claims with that fill_id as invalidated
+    for claim in context.claims:
+        fill_hash_of_claim = request.fill_hash_with_fill_id(claim.latest_claim_made.fill_id)
+        if claim.request_id == request.id and encode_hex(fill_hash_of_claim) == event.fill_hash:
+            claim.invalidate()
+
+    return True, None
```

Comment:
We should return `True` here only if we invalidated any claims, right?

Reply:
It sounds that @istankovic 's question relates more to the fact if the agent or any other agent invalidated the claim. In that case, I'd say it shouldn't matter who invalidated.

Reply:
Hm, I see your concern, but on the other hand, my concern is that if we somehow get events out of order, we will drop the invalidate event and thus never be able to have the correct state for that claim.
So I would rather prefer not to drop such events.

Also, we probably have additional places/events where this might be of concern and I'd rather that we have a general solution, e.g. an expiration time for such events in the sense "drop the event if we are unable to proceed with it for more than N hours/minutes/etc.".

Reply:
Another question which comes to my mind in this discussion. If `FillHashInvalidated` is emitted by the `ResolutionRegistry`, are we also handling the `HashInvalidated` Event emitted by `FillManager`?

Reply:
Good point, for this case we don't need to L1 relay.

Reply:
These two events should somehow be reflected in the state machine I guess. Because with `HashInvalidated` Event an agent would know that he doesn't need to call `invalidateFillHash` on the `FillManager` anymore.